Fix logic to check for empty results in MS/Common

amaltaro opened the pull request:
Fixes #11147 

#### Status
ready

#### Description
Summary of changes is:
* check if the HTTP request was successful based solely on the status code (anything other than code 200 is considered as failed)
* change the exception type
* marked a few unit tests as integration to avoid making calls to real DBS server

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

amaltaro committed:
Fix logic to check for empty results in MS/Common

change check from empty result to http failed

pass HTTP request only for status code 200

amaltaro committed:
unit tests

update unit tests

update unit tests

amaltaro commented:
test this please

vkuznet commented on src/python/WMCore/MicroService/Tools/Common.py:
```diff
@@ -34,22 +34,21 @@
     code = int(row.get('code', 200))
-    data = row['data']
-    if (code >= 200 and code < 400) and data in (None, []):
+    if code >= 400 and 'error' in row:
```
@amaltaro this will not work for non 200 requests since the response will (may) contain empty body. For instance, if HTTP server has moved its API to another end-point, then you'll get 301 HTTP StatusMovedPermanently code. (This is common use-case when server only deal with `/api` but can't handle extra slash like `/api/ and vice versa. Now, the HTTP response body in this case will be empty. And, if we use Python client it will fail later in a code, see:
```
if hasHTTPFailed(row):
   ...
rows = json.loads(row['data'])
...
for item in rows:
```

So, if HTTP response body is empty string then `json.loads` will fail as following:
```
>>> import json
>>> json.loads('')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

So, in addition to checking HTTP code we also need to wrap `rows = json.loads(row['data'])` into try/except block or check that `row['data']` is not equal to empty string. Otherwise, we should only concentrate on 200 HTTP code only. I think to satisfy both use-cases it is better to wrap `rows = json.loads(row['data'])` into try/except block within this module. The except should provide error handling and provide details about code and response body, e.g.
```
try:
        rows = json.loads(row['data'])
        ...
        for item in rows:
except:
        print("HTTP code {}, response {}".format(row.get('code'), row.get('data'))
```

vkuznet commented on test/python/WMCore_t/MicroService_t/Tools_t/Common_t.py:
```diff
@@ -31,36 +35,44 @@ def testDbsInfo(self):
+    def testHasHTTPFailed(self):
+        """Test the hasHTTPFailed method"""
+        self.assertTrue(hasHTTPFailed({}))
+        self.assertTrue(hasHTTPFailed({'data': 1, 'code': 400, 'error': 'blah'}))
```
please add additional test with
```
{''data': '', 'code': 301}
```
and handle it correctly according how code will be adjusted.

amaltaro commented:
Thank you for your suggestion, Valentin. I have changed the code now to mark HTTP requests as successful only when their status code is 200. Unit test was also updated according to your suggestion. Thanks

goughes commented:
test this please
